File: Algorithms/p059_Spiral_Matrix_II/p059_Spiral_Matrix_II.cpp
class Solution {
public:
    vector<vector<int>> generateMatrix(int n) {
        vector<vector<int>> res(n,vector<int>(n));
        int l = 0;
        int r = n-1;
        int u = 0;
        int d = n-1;
        int i = 0;
        while(1){
            if(l>r) break;
            for(int k=l;k<=r;++k){
                res[u][k] = ++i;
            }
            ++u;
            if(u>d) break;
            for(int k=u;k<=d;++k){
                res[k][r] = ++i;
            }
            --r;
            if(l>r) break;
            for(int k=r;k>=l;--k){
                res[d][k] = ++i;
            }
            --d;
            if(u>d) break;
            for(int k=d;k>=u;--k){
                res[k][l] = ++i;
            }
            ++l;
        }
        return res;
    }
};

File: Algorithms/p059_Spiral_Matrix_II/p059_Spiral_Matrix_II.py
class Solution(object):
    def generateMatrix(self, n):
        """
        :type n: int
        :rtype: List[List[int]]
        """
        res = [[None for j in range(0,n)] for i in range(0,n)]
        l = 0
        r = n-1
        u = 0
        d = n-1
        i = 0
        while True:
            if l>r:
                break
            for k in range(l,r+1):
                i += 1
                res[u][k] = i
            u += 1
            if u>d:
                break
            for k in range(u,d+1):
                i += 1
                res[k][r] = i
            r -= 1
            if l>r:
                break
            for k in range(r,l-1,-1):
                i += 1
                res[d][k] = i
            d -= 1
            if u>d:
                break
            for k in range(d,u-1,-1):
                i += 1
                res[k][l] = i
            l += 1
        return res
